File: 2021/Day4/Non overlapping intervals/non_overlapping_intervals.cpp
bool cmpr(vector<int> &a, vector<int> &b)
{
    return a[1] < b[1];
}
class Solution
{
public:
    int eraseOverlapIntervals(vector<vector<int>> &intervals)
    {
        //sort on the basis of end time
        //set of jobs which can give us max non overlapping
        //iterate through each job and then see if prev[1]> current[0] - true - add++
        //otherwise - prev = curr;
        //By CyFun
        int n = intervals.size();
        int ans = -1;
        if (n == 0)
            return 0;
        sort(intervals.begin(), intervals.end(), cmpr);
        vector<int> prev = intervals[0];
        for (auto i : intervals)
        {
            if (prev[1] > i[0])
                ans++;
            else
                prev = i;
        }
        return ans;
    }
};

File: 2021/Day4/Non overlapping intervals/non_overlapping_intervals.py
class Solution:# By CyFun
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        if len(intervals) <= 1:
            return 0
        intervals.sort()
        overleap = 0

        end = intervals[0][1]

        for i in range(1, len(intervals)):
            if intervals[i][0] < end:
                overleap += 1
                end = min(intervals[i][1], end)
            else:
                end = intervals[i][1]

        return overleap
